ceremonies/src/lib.rs

```diff
@@ -173,6 +173,94 @@ pub mod pallet {
 			Ok(().into())
 		}
 
+		#[pallet::weight((<T as Config>::WeightInfo::upgrade_registration(), DispatchClass::Normal, Pays::Yes))]
+		pub fn upgrade_registration(
+			origin: OriginFor<T>,
+			cid: CommunityIdentifier,
+			proof: ProofOfAttendance<T::Signature, T::AccountId>,
+		) -> DispatchResultWithPostInfo {
+			let sender = ensure_signed(origin.clone())?;
+			let current_phase = <encointer_scheduler::Pallet<T>>::current_phase();
+			ensure!(
+				<encointer_communities::Pallet<T>>::community_identifiers().contains(&cid),
+				Error::<T>::InexistentCommunity
+			);
+
+			let mut cindex = <encointer_scheduler::Pallet<T>>::current_ceremony_index();
+
+			if current_phase == CeremonyPhaseType::Attesting {
+				cindex += 1
+			};
+
+			if let Some(participant_type) = Self::get_participant_type((cid, cindex), &sender) {
+				if participant_type == ParticipantType::Newbie {
+					Self::remove_participant_from_registry(cid, cindex, &sender)?;
+					Self::register_participant(origin, cid, Some(proof))?;
+				} else {
+					return Err(<Error<T>>::MustBeNewbieToUpgradeRegistration.into())
+				}
+			} else {
+				return Err(<Error<T>>::ParticipantIsNotRegistered.into())
+			}
+			Ok(().into())
+		}
+
+		#[pallet::weight((<T as Config>::WeightInfo::unregister_participant(), DispatchClass::Normal, Pays::Yes))]
+		pub fn unregister_participant(
+			origin: OriginFor<T>,
+			cid: CommunityIdentifier,
+			maybe_reputation_community_ceremony: Option<CommunityCeremony>,
+		) -> DispatchResultWithPostInfo {
+			let sender = ensure_signed(origin)?;
+			let current_phase = <encointer_scheduler::Pallet<T>>::current_phase();
+			ensure!(
+				vec![CeremonyPhaseType::Registering, CeremonyPhaseType::Attesting]
+					.contains(&current_phase),
+				Error::<T>::RegisteringOrAttestationPhaseRequired
+			);
```

Comment:
1. I dont call it because it will be called later in `register_participant` and so we dont call it twice. are extrinsic calls atomic in substrate? if not, you are right and we need it.
2. I used the vector logic because of the fact that your approach assumes implicit knowledge about `CeremonyPhaseType`. Assume we use your logic and at some point in the future add another phase type, this will break our method. But maybe there is a better way to do it than with a vec?

Reply:
1. Ah I see, I did not get that we call the top level dispatchable. However, transactions are not atomic in the sense that returning an error will result in no state being changed. All state changes that occur before the error is returned will persist. This is why, we must adhere to "verify first - write last" write last paradigm.
2. This is true, but I would argue two things here: Changing the the `CeremonyPhaseType`, will very likely also change the semantics in what we need to do at this step. Secondly, if we wrap the logic into the function, we don't need to care about the implementation details, the maintainer of the `CeremonyPhaseType` must guarantee that the function will do what its name implies.